## Assignment – Sparsity

-by Qi Sun

**1. Jokester**

The dataset that I picked is jester-data-3: (2.1MB) Data from 24,938 users who have rated between 15 and 35 jokes, a matrix with dimensions 24,938 X 101.

* Ratings are real values ranging from -10.00 to +10.00 (the value "99" corresponds to "null" = "not rated").

* One row per user

* The first column gives the number of jokes rated by that user. The next 100 columns give the ratings for jokes 01 - 100.


In [24]:
import pandas as pd
import numpy as np

df = pd.read_csv('https://raw.githubusercontent.com/susanqisun/DAV6300/main/jester-data-3.csv', header=None)
df

0     1      2     3     4     5      6      7     8     9    ...  \
0       26  99.0  99.00  99.0  99.0 -1.65  99.00  -0.78  6.89  99.0  ...   
1       33  99.0  99.00  99.0  99.0 -9.27  99.00  -9.17 -8.59  99.0  ...   
2       16  99.0  99.00  99.0  99.0 -6.12  99.00  -7.48 -7.77  99.0  ...   
3       24  99.0   0.05  99.0  99.0 -2.82  99.00  -4.85 -0.87  99.0  ...   
4       22  99.0  99.00  99.0  99.0 -4.95  99.00   6.21  2.72  99.0  ...   
...    ...   ...    ...   ...   ...   ...    ...    ...   ...   ...  ...   
24933   21  99.0  99.00  99.0  99.0  5.78   8.45  99.00  8.20  99.0  ...   
24934   16  99.0  99.00  99.0  99.0  5.24  99.00  99.00  5.68  99.0  ...   
24935   35  99.0  99.00  99.0  99.0 -3.93  99.00  -9.03 -2.67  99.0  ...   
24936   21  99.0  99.00  99.0  99.0  9.42   9.66  99.00  9.17  99.0  ...   
24937   35  99.0  99.00  99.0  99.0  0.19  99.00  -1.50  7.82  99.0  ...   

        91    92     93    94     95    96    97    98     99    100  
0      99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
1      99.0  99.0  -2.77  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
2      99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
3      99.0  99.0  99.00  99.0   1.84  99.0  99.0  99.0  -4.08  99.0  
4      99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
...     ...   ...    ...   ...    ...   ...   ...   ...    ...   ...  
24933  99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
24934  99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
24935  99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
24936  99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
24937  99.0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  

[24938 rows x 101 columns]

In [25]:
# check the shape for the dataframe
print(df.shape)

(24938, 101)


In [26]:
#check missing value
df.isnull().sum()

0      0
1      0
2      0
3      0
4      0
      ..
96     0
97     0
98     0
99     0
100    0
Length: 101, dtype: int64

### Data Preprocessing

Dataset contains no column names. The first column gives the number of jokes rated by that user. The next 100 columns give the ratings for jokes 01 - 100. 


#### 1. Rename column

In [27]:
#0th column would be renamed to freq
df.rename(columns = {0: 'freq'}, inplace = True)
df.head()

freq     1      2     3     4     5     6     7     8     9  ...    91  \
0    26  99.0  99.00  99.0  99.0 -1.65  99.0 -0.78  6.89  99.0  ...  99.0   
1    33  99.0  99.00  99.0  99.0 -9.27  99.0 -9.17 -8.59  99.0  ...  99.0   
2    16  99.0  99.00  99.0  99.0 -6.12  99.0 -7.48 -7.77  99.0  ...  99.0   
3    24  99.0   0.05  99.0  99.0 -2.82  99.0 -4.85 -0.87  99.0  ...  99.0   
4    22  99.0  99.00  99.0  99.0 -4.95  99.0  6.21  2.72  99.0  ...  99.0   

     92     93    94     95    96    97    98     99   100  
0  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
1  99.0  -2.77  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
2  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  
3  99.0  99.00  99.0   1.84  99.0  99.0  99.0  -4.08  99.0  
4  99.0  99.00  99.0  99.00  99.0  99.0  99.0  99.00  99.0  

[5 rows x 101 columns]

#### 2. Replace all 99.0 ratings as 0

The value "99" corresponds to "null" = "not rated", which means jokes were not rated by users. I'll replace them as 0.

In [28]:
df02 = df.replace(99.0, 0)
df02.head()

freq    1     2    3    4     5    6     7     8    9  ...   91   92    93  \
0    26  0.0  0.00  0.0  0.0 -1.65  0.0 -0.78  6.89  0.0  ...  0.0  0.0  0.00   
1    33  0.0  0.00  0.0  0.0 -9.27  0.0 -9.17 -8.59  0.0  ...  0.0  0.0 -2.77   
2    16  0.0  0.00  0.0  0.0 -6.12  0.0 -7.48 -7.77  0.0  ...  0.0  0.0  0.00   
3    24  0.0  0.05  0.0  0.0 -2.82  0.0 -4.85 -0.87  0.0  ...  0.0  0.0  0.00   
4    22  0.0  0.00  0.0  0.0 -4.95  0.0  6.21  2.72  0.0  ...  0.0  0.0  0.00   

    94    95   96   97   98    99  100  
0  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
1  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
2  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
3  0.0  1.84  0.0  0.0  0.0 -4.08  0.0  
4  0.0  0.00  0.0  0.0  0.0  0.00  0.0  

[5 rows x 101 columns]

#### 3. Transform from wide to long

In [34]:
df03 = df02.copy()
df03.index = np.arange(1, len(df02) + 1)
df03

freq    1     2    3    4     5     6     7     8    9  ...   91   92  \
1        26  0.0  0.00  0.0  0.0 -1.65  0.00 -0.78  6.89  0.0  ...  0.0  0.0   
2        33  0.0  0.00  0.0  0.0 -9.27  0.00 -9.17 -8.59  0.0  ...  0.0  0.0   
3        16  0.0  0.00  0.0  0.0 -6.12  0.00 -7.48 -7.77  0.0  ...  0.0  0.0   
4        24  0.0  0.05  0.0  0.0 -2.82  0.00 -4.85 -0.87  0.0  ...  0.0  0.0   
5        22  0.0  0.00  0.0  0.0 -4.95  0.00  6.21  2.72  0.0  ...  0.0  0.0   
...     ...  ...   ...  ...  ...   ...   ...   ...   ...  ...  ...  ...  ...   
24934    21  0.0  0.00  0.0  0.0  5.78  8.45  0.00  8.20  0.0  ...  0.0  0.0   
24935    16  0.0  0.00  0.0  0.0  5.24  0.00  0.00  5.68  0.0  ...  0.0  0.0   
24936    35  0.0  0.00  0.0  0.0 -3.93  0.00 -9.03 -2.67  0.0  ...  0.0  0.0   
24937    21  0.0  0.00  0.0  0.0  9.42  9.66  0.00  9.17  0.0  ...  0.0  0.0   
24938    35  0.0  0.00  0.0  0.0  0.19  0.00 -1.50  7.82  0.0  ...  0.0  0.0   

         93   94    95   96   97   98    99  100  
1      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
2     -2.77  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
3      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
4      0.00  0.0  1.84  0.0  0.0  0.0 -4.08  0.0  
5      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
...     ...  ...   ...  ...  ...  ...   ...  ...  
24934  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24935  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24936  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24937  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24938  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  

[24938 rows x 101 columns]

In [44]:
# Save the 'freq' coloumn that I don't want to scale
freq = df03['freq'].reset_index()

freq.rename(columns = {'index': 'UserId'}, inplace = True)

freq

UserId  freq
0           1    26
1           2    33
2           3    16
3           4    24
4           5    22
...       ...   ...
24933   24934    21
24934   24935    16
24935   24936    35
24936   24937    21
24937   24938    35

[24938 rows x 2 columns]

In [36]:
#Extract ratings columns
ratings = df03.iloc[:, 1:]
ratings

1     2    3    4     5     6     7     8    9     10   ...  91   92   \
1      0.0  0.00  0.0  0.0 -1.65  0.00 -0.78  6.89  0.0  0.00  ...  0.0  0.0   
2      0.0  0.00  0.0  0.0 -9.27  0.00 -9.17 -8.59  0.0  0.00  ...  0.0  0.0   
3      0.0  0.00  0.0  0.0 -6.12  0.00 -7.48 -7.77  0.0  0.00  ...  0.0  0.0   
4      0.0  0.05  0.0  0.0 -2.82  0.00 -4.85 -0.87  0.0  0.00  ...  0.0  0.0   
5      0.0  0.00  0.0  0.0 -4.95  0.00  6.21  2.72  0.0 -8.59  ...  0.0  0.0   
...    ...   ...  ...  ...   ...   ...   ...   ...  ...   ...  ...  ...  ...   
24934  0.0  0.00  0.0  0.0  5.78  8.45  0.00  8.20  0.0  8.50  ...  0.0  0.0   
24935  0.0  0.00  0.0  0.0  5.24  0.00  0.00  5.68  0.0  0.00  ...  0.0  0.0   
24936  0.0  0.00  0.0  0.0 -3.93  0.00 -9.03 -2.67  0.0  0.00  ...  0.0  0.0   
24937  0.0  0.00  0.0  0.0  9.42  9.66  0.00  9.17  0.0  0.00  ...  0.0  0.0   
24938  0.0  0.00  0.0  0.0  0.19  0.00 -1.50  7.82  0.0  0.00  ...  0.0  0.0   

        93   94    95   96   97   98    99   100  
1      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
2     -2.77  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
3      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
4      0.00  0.0  1.84  0.0  0.0  0.0 -4.08  0.0  
5      0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
...     ...  ...   ...  ...  ...  ...   ...  ...  
24934  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24935  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24936  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24937  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  
24938  0.00  0.0  0.00  0.0  0.0  0.0  0.00  0.0  

[24938 rows x 100 columns]

In [38]:
df04 = ratings.stack().reset_index()
df04.columns=['UserId', 'jokeID', 'rating']

df04

UserId  jokeID  rating
0             1       1    0.00
1             1       2    0.00
2             1       3    0.00
3             1       4    0.00
4             1       5   -1.65
...         ...     ...     ...
2493795   24938      96    0.00
2493796   24938      97    0.00
2493797   24938      98    0.00
2493798   24938      99    0.00
2493799   24938     100    0.00

[2493800 rows x 3 columns]

In [72]:
# add freq column
# merge together
df05 = pd.merge(left=freq, right=df04, how='outer')
df05

UserId  freq  jokeID  rating
0             1    26       1    0.00
1             1    26       2    0.00
2             1    26       3    0.00
3             1    26       4    0.00
4             1    26       5   -1.65
...         ...   ...     ...     ...
2493795   24938    35      96    0.00
2493796   24938    35      97    0.00
2493797   24938    35      98    0.00
2493798   24938    35      99    0.00
2493799   24938    35     100    0.00

[2493800 rows x 4 columns]

#### Sparsity of a matrix:

Sparsity of Matrix = Number of Empty cells / Total Number of cells

In [74]:
(df05['rating']==0).sum() / 2493800

0.7533519127436041

It means that 75.34% of cells are empty. 

#### 4. Normalize ratings

In this dataset, there are positive and negative ratings. For example, for jokes 01, the lowest rating is -9.66, and the highest rating is 8.88. I'll normalize ratings columns by using StandardScalar, so it's distribution will have mean of 0 and standard deviation of 1. 

In [9]:
ratings.describe()

1             2             3             4             5    \
count  24937.000000  24937.000000  24937.000000  24937.000000  24937.000000   
mean       0.005181      0.007449      0.000302      0.000220      0.180777   
std        0.283392      0.426281      0.069716      0.025294      5.378339   
min       -9.660000     -9.810000     -3.980000      0.000000     -9.950000   
25%        0.000000      0.000000      0.000000      0.000000     -4.270000   
50%        0.000000      0.000000      0.000000      0.000000      0.830000   
75%        0.000000      0.000000      0.000000      0.000000      4.470000   
max        8.880000      9.170000      9.080000      3.790000      9.510000   

                6             7             8             9             10   \
count  24937.000000  24937.000000  24937.000000  24937.000000  24937.000000   
mean       0.002277     -0.971692     -1.631795     -0.000072      0.009153   
std        0.153461      5.718898      4.976294      0.011399      0.388179   
min       -8.250000     -9.950000     -9.950000     -1.800000     -9.810000   
25%        0.000000     -6.410000     -6.120000      0.000000      0.000000   
50%        0.000000     -0.580000     -1.600000      0.000000      0.000000   
75%        0.000000      3.690000      2.180000      0.000000      0.000000   
max        9.660000     10.000000      9.760000      0.000000      9.320000   

       ...           91            92            93            94   \
count  ...  24937.000000  24937.000000  24937.000000  24937.000000   
mean   ...      0.043991      0.011256      0.088053     -0.011158   
std    ...      1.539790      1.460806      1.466384      1.449652   
min    ...     -9.950000     -9.950000     -9.950000     -9.950000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      0.000000      0.000000      0.000000      0.000000   
max    ...      9.370000      9.610000      9.370000      9.370000   

                95            96            97            98            99   \
count  24937.000000  24937.000000  24937.000000  24937.000000  24937.000000   
mean       0.017064      0.014542      0.007350      0.027865     -0.056094   
std        1.474790      1.401022      1.473256      1.461787      1.513393   
min       -9.950000     -9.950000     -9.950000     -9.950000     -9.950000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        9.370000      9.370000      9.370000      9.370000      9.810000   

                100  
count  24937.000000  
mean       0.000095  
std        0.090474  
min       -8.790000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        9.220000  

[8 rows x 100 columns]

In [48]:
#Fit StandardScaler into ratings
from sklearn.preprocessing import StandardScaler

df05[['rating']] = StandardScaler().fit_transform(df05[['rating']])
df05

UserId  freq  jokeID    rating
0             1    26       1 -0.027119
1             1    26       2 -0.027119
2             1    26       3 -0.027119
3             1    26       4 -0.027119
4             1    26       5 -0.637410
...         ...   ...     ...       ...
2493795   24938    35      96 -0.027119
2493796   24938    35      97 -0.027119
2493797   24938    35      98 -0.027119
2493798   24938    35      99 -0.027119
2493799   24938    35     100 -0.027119

[2493800 rows x 4 columns]

------------------

**2. MovieLens**

The dataset that I used is MovieLens 100K Dataset (u.data).

In [50]:
# read rating dataset
column_names = ['user_id', 'item_id', 'rating','timestamp']
df11 = pd.read_csv('https://raw.githubusercontent.com/susanqisun/DAV6300/main/u.data.txt', sep='\t', names=column_names)
df11


user_id  item_id  rating  timestamp
0          196      242       3  881250949
1          186      302       3  891717742
2           22      377       1  878887116
3          244       51       2  880606923
4          166      346       1  886397596
...        ...      ...     ...        ...
99995      880      476       3  880175444
99996      716      204       5  879795543
99997      276     1090       1  874795795
99998       13      225       2  882399156
99999       12      203       3  879959583

[100000 rows x 4 columns]

In [51]:
df11.shape

(100000, 4)

In [56]:
# read the movie titles
movies = pd.read_csv('https://raw.githubusercontent.com/susanqisun/DAV6300/main/u.item.txt', sep='\|', header=None,encoding = "ISO-8859-1")

movies

/Users/yangyang/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:2: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  


0                                          1            2   3   \
0        1                           Toy Story (1995)  01-Jan-1995 NaN   
1        2                           GoldenEye (1995)  01-Jan-1995 NaN   
2        3                          Four Rooms (1995)  01-Jan-1995 NaN   
3        4                          Get Shorty (1995)  01-Jan-1995 NaN   
4        5                             Copycat (1995)  01-Jan-1995 NaN   
...    ...                                        ...          ...  ..   
1677  1678                          Mat' i syn (1997)  06-Feb-1998 NaN   
1678  1679                           B. Monkey (1998)  06-Feb-1998 NaN   
1679  1680                       Sliding Doors (1998)  01-Jan-1998 NaN   
1680  1681                        You So Crazy (1994)  01-Jan-1994 NaN   
1681  1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996 NaN   

                                                     4   5   6   7   8   9   \
0     http://us.imdb.com/M/title-exact?Toy%20Story%2...   0   0   0   1   1   
1     http://us.imdb.com/M/title-exact?GoldenEye%20(...   0   1   1   0   0   
2     http://us.imdb.com/M/title-exact?Four%20Rooms%...   0   0   0   0   0   
3     http://us.imdb.com/M/title-exact?Get%20Shorty%...   0   1   0   0   0   
4     http://us.imdb.com/M/title-exact?Copycat%20(1995)   0   0   0   0   0   
...                                                 ...  ..  ..  ..  ..  ..   
1677  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...   0   0   0   0   0   
1678  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...   0   0   0   0   0   
1679      http://us.imdb.com/Title?Sliding+Doors+(1998)   0   0   0   0   0   
1680  http://us.imdb.com/M/title-exact?You%20So%20Cr...   0   0   0   0   0   
1681  http://us.imdb.com/M/title-exact?Schrei%20aus%...   0   0   0   0   0   

      ...  14  15  16  17  18  19  20  21  22  23  
0     ...   0   0   0   0   0   0   0   0   0   0  
1     ...   0   0   0   0   0   0   0   1   0   0  
2     ...   0   0   0   0   0   0   0   1   0   0  
3     ...   0   0   0   0   0   0   0   0   0   0  
4     ...   0   0   0   0   0   0   0   1   0   0  
...   ...  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  
1677  ...   0   0   0   0   0   0   0   0   0   0  
1678  ...   0   0   0   0   0   1   0   1   0   0  
1679  ...   0   0   0   0   0   1   0   0   0   0  
1680  ...   0   0   0   0   0   0   0   0   0   0  
1681  ...   0   0   0   0   0   0   0   0   0   0  

[1682 rows x 24 columns]

In [57]:
movies.shape

(1682, 24)

In [58]:
movies_title = movies[[0,1]]
movies_title

0                                          1
0        1                           Toy Story (1995)
1        2                           GoldenEye (1995)
2        3                          Four Rooms (1995)
3        4                          Get Shorty (1995)
4        5                             Copycat (1995)
...    ...                                        ...
1677  1678                          Mat' i syn (1997)
1678  1679                           B. Monkey (1998)
1679  1680                       Sliding Doors (1998)
1680  1681                        You So Crazy (1994)
1681  1682  Scream of Stone (Schrei aus Stein) (1991)

[1682 rows x 2 columns]

In [59]:
movies_title.columns = ['item_id', 'title']
movies_title

item_id                                      title
0           1                           Toy Story (1995)
1           2                           GoldenEye (1995)
2           3                          Four Rooms (1995)
3           4                          Get Shorty (1995)
4           5                             Copycat (1995)
...       ...                                        ...
1677     1678                          Mat' i syn (1997)
1678     1679                           B. Monkey (1998)
1679     1680                       Sliding Doors (1998)
1680     1681                        You So Crazy (1994)
1681     1682  Scream of Stone (Schrei aus Stein) (1991)

[1682 rows x 2 columns]

In [60]:
# Merging together
df12 = pd.merge(df11, movies_title, on='item_id')
df12

user_id  item_id  rating  timestamp  \
0          196      242       3  881250949   
1           63      242       3  875747190   
2          226      242       5  883888671   
3          154      242       3  879138235   
4          306      242       5  876503793   
...        ...      ...     ...        ...   
99995      840     1674       4  891211682   
99996      655     1640       3  888474646   
99997      655     1637       3  888984255   
99998      655     1630       3  887428735   
99999      655     1641       3  887427810   

                                                   title  
0                                           Kolya (1996)  
1                                           Kolya (1996)  
2                                           Kolya (1996)  
3                                           Kolya (1996)  
4                                           Kolya (1996)  
...                                                  ...  
99995                                  Mamma Roma (1962)  
99996                             Eighth Day, The (1996)  
99997                                  Girls Town (1996)  
99998  Silence of the Palace, The (Saimt el Qusur) (1...  
99999                                    Dadetown (1995)  

[100000 rows x 5 columns]

#### Sparsity of a matrix:

unique users: 943

unique movies: 1682

In [85]:
# To check the number of unique users
df11['user_id'].nunique()

943

In [86]:
# Checking the number of unique movies
df11['item_id'].nunique()

1682

In [78]:
((943*1682)-100000) / (943*1682)

0.9369533063577546

#### How many times a movie has been watched?

In [79]:
# How many times a movie has been watched
df12.groupby('title').count()['rating'].sort_values(ascending=False)

title
Star Wars (1977)                              583
Contact (1997)                                509
Fargo (1996)                                  508
Return of the Jedi (1983)                     507
Liar Liar (1997)                              485
                                             ... 
Man from Down Under, The (1943)                 1
Marlene Dietrich: Shadow and Light (1996)       1
Mat' i syn (1997)                               1
Mille bolle blu (1993)                          1
Á köldum klaka (Cold Fever) (1994)              1
Name: rating, Length: 1664, dtype: int64

#### Calculate average rating for each movie

In [67]:
# Finding the average rating of a movie
df12.groupby(['title'], as_index = False, sort = False).mean()[['title','rating']]

title    rating
0                                          Kolya (1996)  3.991453
1                              L.A. Confidential (1997)  4.161616
2                                   Heavyweights (1994)  2.153846
3                            Legends of the Fall (1994)  3.456790
4                                   Jackie Brown (1997)  3.642857
...                                                 ...       ...
1659                                  Mamma Roma (1962)  4.000000
1660                             Eighth Day, The (1996)  3.000000
1661                                  Girls Town (1996)  3.000000
1662  Silence of the Palace, The (Saimt el Qusur) (1...  3.000000
1663                                    Dadetown (1995)  3.000000

[1664 rows x 2 columns]

--------------
**3. Movie Survey**

In [80]:
# survey results
df21 = pd.read_csv('https://raw.githubusercontent.com/susanqisun/DAV6300/main/movie%20recommender.csv')
df21

UserID The Little Things The White Tiger   The Dig      Soul  \
0       1                 3               4         5  Not Sure   
1       2          Not Sure        Not Sure         4         5   
2       3                 5               4         5         5   
3       4          Not Sure        Not Sure  Not Sure         3   
4       5                 5               5         4         3   
5       6                 3               2       NaN         3   
6       7               NaN               1       NaN       NaN   

   Wonder Woman 1984  Promising Young Woman  
0                  1                    3.0  
1                  5                    4.0  
2                  3                    4.0  
3                  4                    5.0  
4                  4                    2.0  
5                  4                    5.0  
6                  4                    NaN

In [81]:
df22 = df21.replace('Not Sure',np.NaN)
df22

UserID The Little Things The White Tiger The Dig Soul  Wonder Woman 1984  \
0       1                 3               4       5  NaN                  1   
1       2               NaN             NaN       4    5                  5   
2       3                 5               4       5    5                  3   
3       4               NaN             NaN     NaN    3                  4   
4       5                 5               5       4    3                  4   
5       6                 3               2     NaN    3                  4   
6       7               NaN               1     NaN  NaN                  4   

   Promising Young Woman  
0                    3.0  
1                    4.0  
2                    4.0  
3                    5.0  
4                    2.0  
5                    5.0  
6                    NaN

In [83]:
df23 = df22.drop(['UserID'], axis=1)

df_survey = df23.stack().reset_index()
df_survey.columns=['UserId', 'movie', 'rating']

df_survey

UserId                  movie rating
0        0      The Little Things      3
1        0        The White Tiger      4
2        0                The Dig      5
3        0      Wonder Woman 1984      1
4        0  Promising Young Woman      3
5        1                The Dig      4
6        1                   Soul      5
7        1      Wonder Woman 1984      5
8        1  Promising Young Woman      4
9        2      The Little Things      5
10       2        The White Tiger      4
11       2                The Dig      5
12       2                   Soul      5
13       2      Wonder Woman 1984      3
14       2  Promising Young Woman      4
15       3                   Soul      3
16       3      Wonder Woman 1984      4
17       3  Promising Young Woman      5
18       4      The Little Things      5
19       4        The White Tiger      5
20       4                The Dig      4
21       4                   Soul      3
22       4      Wonder Woman 1984      4
23       4  Promising Young Woman      2
24       5      The Little Things      3
25       5        The White Tiger      2
26       5                   Soul      3
27       5      Wonder Woman 1984      4
28       5  Promising Young Woman      5
29       6        The White Tiger      1
30       6      Wonder Woman 1984      4

#### Sparsity of a matrix:


In [84]:
31/42

0.7380952380952381

--------------------
**4. Book-Crossing Dataset**

Book-Crossings is a book ratings dataset compiled by Cai-Nicolas Ziegler. The ratings are on a scale from 1 to 10. The data consists of three tables: ratings, books info, and users info.

http://www2.informatik.uni-freiburg.de/~cziegler/BX/

Since the file size is larger than 25MB, I'll upload datasets from my local folder.

In [93]:
books = pd.read_csv('/Users/yangyang/Downloads/BX-CSV-Dump/BX-Books.csv', sep=';', error_bad_lines=False, encoding="latin-1")
books.columns = ['ISBN', 'bookTitle', 'bookAuthor', 'yearOfPublication', 'publisher', 'imageUrlS', 'imageUrlM', 'imageUrlL']
users = pd.read_csv('/Users/yangyang/Downloads/BX-CSV-Dump/BX-Users.csv', sep=';', error_bad_lines=False, encoding="latin-1")
users.columns = ['userID', 'Location', 'Age']
ratings = pd.read_csv('/Users/yangyang/Downloads/BX-CSV-Dump/BX-Book-Ratings.csv', sep=';', error_bad_lines=False, encoding="latin-1")
ratings.columns = ['userID', 'ISBN', 'bookRating']

b'Skipping line 6452: expected 8 fields, saw 9\nSkipping line 43667: expected 8 fields, saw 10\nSkipping line 51751: expected 8 fields, saw 9\n'
b'Skipping line 92038: expected 8 fields, saw 9\nSkipping line 104319: expected 8 fields, saw 9\nSkipping line 121768: expected 8 fields, saw 9\n'
b'Skipping line 144058: expected 8 fields, saw 9\nSkipping line 150789: expected 8 fields, saw 9\nSkipping line 157128: expected 8 fields, saw 9\nSkipping line 180189: expected 8 fields, saw 9\nSkipping line 185738: expected 8 fields, saw 9\n'
b'Skipping line 209388: expected 8 fields, saw 9\nSkipping line 220626: expected 8 fields, saw 9\nSkipping line 227933: expected 8 fields, saw 11\nSkipping line 228957: expected 8 fields, saw 10\nSkipping line 245933: expected 8 fields, saw 9\nSkipping line 251296: expected 8 fields, saw 9\nSkipping line 259941: expected 8 fields, saw 9\nSkipping line 261529: expected 8 fields, saw 9\n'
/Users/yangyang/opt/anaconda3/lib/python3.7/site-packages/IPython/core/int

In [95]:
books.head()

ISBN                                          bookTitle  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

             bookAuthor yearOfPublication                   publisher  \
0    Mark P. O. Morford              2002     Oxford University Press   
1  Richard Bruce Wright              2001       HarperFlamingo Canada   
2          Carlo D'Este              1991             HarperPerennial   
3      Gina Bari Kolata              1999        Farrar Straus Giroux   
4       E. J. W. Barber              1999  W. W. Norton &amp; Company   

                                           imageUrlS  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlM  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                           imageUrlL  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [99]:
# unique books
books['ISBN'].nunique()

271360

In [96]:
users.head()

userID                            Location   Age
0       1                  nyc, new york, usa   NaN
1       2           stockton, california, usa  18.0
2       3     moscow, yukon territory, russia   NaN
3       4           porto, v.n.gaia, portugal  17.0
4       5  farnborough, hants, united kingdom   NaN

In [101]:
(books['ISBN'].nunique())*(users['userID'].nunique())

75670906880

In [100]:
# unique books
users['userID'].nunique()

278858

In [97]:
ratings

userID         ISBN  bookRating
0        276725   034545104X           0
1        276726   0155061224           5
2        276727   0446520802           0
3        276729   052165615X           3
4        276729   0521795028           6
...         ...          ...         ...
1149775  276704   1563526298           9
1149776  276706   0679447156           0
1149777  276709   0515107662          10
1149778  276721   0590442449          10
1149779  276723  05162443314           8

[1149780 rows x 3 columns]

#### Sparsity of a matrix:

Sparsity of Matrix = Number of Empty cells / Total Number of cells

The ratings are on a scale from 1 to 10, so the value of 0 can be viewed as the missing value.


In [107]:
1 - ((1149780 - (ratings['bookRating']==0).sum()) / ((ratings['ISBN'].nunique())*(ratings['userID'].nunique())))

0.9999879047849692